compaction/compaction_manager.cc

```diff
@@ -699,9 +704,15 @@ sstables_task_executor::~sstables_task_executor() {
     _cm._stats.pending_tasks -= _sstables.size() - (_state == state::pending);
 }
 
+void compaction_task_executor::fail_exceptionally(std::exception_ptr ex) noexcept {
+    if (!_compaction_done.available()) {
```

Comment:
How can this be "false"? If the `_compaction_done` is available it means that not only the `forward_to(_compaction_done_promise)` below worked, but also was resolved. But if `forward_to()` worked, then the `perform_task()` above couldn't throw and couldn't call this helper

Reply:
If I understand properly `forward_to` works also for exceptional futures.

So if `do_run()` throws, then _compaction_done is available (and failed). In this situation `perform_task` throws and so we set _compaction_done_promise once again.

Reply:
Ah, I see. Well, this still scares me. If do_run() actually runs, but perform_task() throws afterwards, then it will set_exception on a moved-from promise. I guess it doesn't happen now, but isn't it a time-bomb?

Reply:
I don't understand. If do_run() runs, then `_compaction_done_promise`'s value (or exception) is set and thus `_compaction_done` is available, so we won't enter `if` branch and we're safe

Reply:
Right after `forward_to` we wait until compaction is done
https://github.com/scylladb/scylladb/blob/074db32c4e1d2b79eb37a42ff64d366d1aee4139/compaction/compaction_manager.cc#L715-L716
Isn't it enough?

Reply:
It is, but this link too not-obvious. However, `perform_task()` is private, so probably it's OK